### PR TITLE
DOCS Documentation cleanup

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

On top of #1861

- Cleanup documentation
- Add documentation about using `pyfetch`, `pyimport`

### CODE REVIEW

File: docs/index.rst [hunk 1]
```diff
@@ -7,11 +7,11 @@ Pyodide is a Python distribution for the browser and Node.js based on WebAssembl
 What is Pyodide?
 ----------------
 
-Pyodide is a port of CPython to WebAssembly/[Emscripten](https://emscripten.org/).
+Pyodide is a port of CPython to WebAssembly/`Emscripten <https://emscripten.org/>`.
```
Comment (hoodmane) [inline docs/index.rst:10]:
```suggestion
Pyodide is a port of CPython to WebAssembly/`Emscripten <https://emscripten.org/>`_.
```

File: docs/index.rst [hunk 2]
```diff
@@ -7,11 +7,11 @@ Pyodide is a Python distribution for the browser and Node.js based on WebAssembl
...
 
 Pyodide makes it possible to install and run Python packages in the browser with
-[micropip](https://pyodide.org/en/stable/usage/api/micropip-api.html). Any pure
-Python package with a wheel available on PyPi is supported. Many packages with C
+`micropip <https://pyodide.org/en/stable/usage/api/micropip-api.html>`. Any pure
```
Comment (hoodmane) [inline docs/index.rst:13]:
```suggestion
`micropip <https://pyodide.org/en/stable/usage/api/micropip-api.html>`_. Any pure
```

File: docs/project/about.md [hunk 3]
```diff
@@ -1,12 +1,10 @@
...
-
-Pyodide is a port of CPython to WebAssembly/[Emscripten](https://emscripten.org/).
+Pyodide is a Python distribution for the browser and Node.js based on WebAssembly/[Emscripten](https://emscripten.org/).
 
 Pyodide makes it possible to install and run Python packages in the browser with
 [micropip](https://pyodide.org/en/stable/usage/api/micropip-api.html). Any pure
-Python package with a wheel available on PyPi is supported. Many packages with C
+Python package with a wheel available on PyPI is supported. Many packages with C
 extensions have also been ported for use with Pyodide. These include many
 general-purpose packages such as regex, pyyaml, lxml and scientific Python
```
Comment (hoodmane) [inline docs/project/about.md:9]:
```suggestion
general-purpose packages such as regex, PyYAML, lxml and scientific Python
```

File: docs/project/about.md [hunk 4]
```diff
@@ -1,12 +1,10 @@
...
 packages including numpy, pandas, scipy, matplotlib, and scikit-learn.
```
Comment (hoodmane) [inline docs/project/about.md:10]:
```suggestion
packages including NumPy, pandas, SciPy, Matplotlib, and scikit-learn.
```

File: docs/usage/modularizing.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
+
+(modularizing-python-codes)=
+# Modularizing Python codes
```
Comment (hoodmane) [inline docs/usage/modularizing.md:3]: Maybe "Packaging Python code for Pyodide"?

File: docs/usage/modularizing.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+Pyodide provides a simple API {any}`pyodide.runPython` to run Python codes.
```
Comment (hoodmane) [inline docs/usage/modularizing.md:5]: code is a [mass noun](https://en.wikipedia.org/wiki/Mass_noun)
```suggestion
Pyodide provides a simple API {any}`pyodide.runPython` to run Python code.
```

File: docs/usage/faq.md [hunk 7]
```diff
@@ -2,26 +2,44 @@
 
 ## How can I load external files in Pyodide?
 
-Two possible solutions are,
+In order to use external files in Pyodide, you should download and save them
+to the [Emscripten virtual file system](https://emscripten.org/docs/api_reference/Filesystem-API.html).
```
Comment (hoodmane) [inline docs/usage/faq.md:6]: I think this link is more likely to be confusing than helpful for most people.

File: docs/usage/modularizing.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+await pyodide.runPythonAsync(`
+    from pyodide.http import pyfetch
+    response = await pyfetch("https://.../your_package.tar.gz")
+    await response.unpack_archive()
+`)
+pkg = pyodide.pyimport("your_package");
+pkg.do_something();
+```
+
+```{admonition} How can I prefetch package files before loading Pyodide
```
Comment (hoodmane) [inline docs/usage/modularizing.md:63]: Perhaps we ought to provide a way to wrap Js fetch responses into Python fetch responses too...

File: docs/usage/modularizing.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+However, when your Python codes grow bigger, putting hundreds of lines inside `runPython` is not scalable.
+
+The best way to run Python code with Pyodide is:
+
+1. write a Python package
+1. load your Python package into the Pyodide (Emscripten) virtual file system
+1. import the package with ``let mypkg = pyodide.pyimport("mypkgname")``
+1. call into your package with ``mypkg.some_api(some_args)``.
+
+## How to load Python package into the virtual file system
```
Comment (hoodmane) [inline docs/usage/modularizing.md:15]:
```suggestion
## How to load a Python package into the virtual file system
```

File: docs/usage/modularizing.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+If the package is built as a `wheel` file, use {any}`micropip.install` to
+install the package.
```
Comment (hoodmane) [inline docs/usage/modularizing.md:18]: Worth noting that you can use `unpackArchive` to install it just as if it were a zip file. We should say that the advantage of micropip is that it does dependency resolution.

File: docs/usage/modularizing.md (hunk 9 above)
Comment (hoodmane) [inline docs/usage/modularizing.md:15]: I added a simple example of this here:
https://github.com/pyodide/pyodide-examples/
I think it'd be good to encourage people to use a dev server that zips stuff on the fly like what I put there:
https://github.com/pyodide/pyodide-examples/blob/main/server.py

File: docs/usage/type-conversions.md [hunk 11]
```diff
@@ -58,7 +58,7 @@ of the round trip conversion is the original object (in the sense that they live
 at the same memory address). There are a few exceptions:
 
 1. `NaN` is converted to `NaN` after a round trip but `NaN !== NaN`,
-2. `null` is converted to `undefined` after a round trip, and
+2. `null` is converted to `undefined` after a round trip
```
Comment (hoodmane) [inline docs/usage/type-conversions.md:61]: it's generally considered to be good style to write numbered lists so that joining the entries together makes a grammatically correct sentence
```suggestion
2. `null` is converted to `undefined` after a round trip, and
```

File: docs/development/contributing.md [hunk 12]
```diff
@@ -36,7 +36,7 @@ always passing. The unfortunate reality of software development is sometimes
 things break. As such, `main` cannot be expected to remain reliable at all
 times. We recommend using the latest stable version of Pyodide.
 
-Pyodide follows semantic versioning (http://semver.org/) - major versions for
+Pyodide follows [semantic versioning](http://semver.org/) - major versions for
```
Comment (hoodmane) [inline docs/development/contributing.md:39]: I wonder if we should put some caveat here about being in version `0.x.x`. In actuality, we mostly release any sort of change as `0.x.0` and on a rare occassion release minor changes that people want sooner to `0.0.x`.

File: docs/index.rst [hunk 13]
```diff
@@ -23,12 +23,12 @@ language, catch it in the other), async/await, and much more.
...
 When used inside a browser, Python has full access to the Web APIs.
 
-Try Pyodide (no installation needed)
+Try Pyodide
 ------------------------------------
 
 Try Pyodide in a
 `REPL <https://pyodide.org/en/stable/console.html>`_ directly in
-your browser.
+your browser. (no installation needed)
```
Comment (hoodmane) [inline docs/index.rst:31]:
```suggestion
your browser (no installation needed).
```

**Review by hoodmane — COMMENTED** [review]
Thanks @ryanking13!

File: docs/usage/type-conversions.md (hunk 11 above)
Comment (ryanking13) [reply docs/usage/type-conversions.md:61]: Thanks! I've learned one English writing style today 😃

File: docs/usage/modularizing.md (hunk 5 above)
Comment (ryanking13) [reply docs/usage/modularizing.md:3]: Yes, that seems better.

File: docs/development/contributing.md (hunk 12 above)
Comment (ryanking13) [reply docs/development/contributing.md:39]: Sorry, I don't understand the meaning clearly.

File: docs/usage/faq.md (hunk 7 above)
Comment (ryanking13) [reply docs/usage/faq.md:6]: Yes, I agree.

File: docs/usage/modularizing.md (hunk 5 above)
Comment (ryanking13) [reply docs/usage/modularizing.md:3]: Hmm... but someone might be confused about the difference between "Creating a Pyodide package" and "Packaging Python code for Pyodide"

File: docs/usage/modularizing.md (hunk 9 above)
Comment (ryanking13) [reply docs/usage/modularizing.md]: One step further, I think it would be nice if we provide a way to integrate the development process with `webpack devserver` of `rollup devserver`

File: docs/usage/modularizing.md (hunk 10 above)
Comment (ryanking13) [reply docs/usage/modularizing.md:18]: Added some warning at the bottom of the document

File: docs/usage/modularizing.md (hunk 5 above)
Comment (hoodmane) [reply docs/usage/modularizing.md:3]: Bundling Python code for Pyodide?

File: docs/usage/faq.md [hunk 14]
```diff
@@ -2,26 +2,44 @@
...
-For security reasons JavaScript in the browser is not allowed to load local data
-files. You need to serve them with a web-browser. There is a
+For security reasons JavaScript in the browser is not allowed to load local data files
+(for example, `file:///path/to/local/file.data`).
+You will run into Network Errors, due to the [Same Origin Policy](https://en.wikipedia.org/wiki/Same-origin_policy).
+There is a
 [File System API](https://wicg.github.io/file-system-access/) supported in Chrome
 but not in Firefox or Safari.
 
+For development purposes, you can serve your files in a
```
Comment (rth) [inline docs/usage/faq.md:39]:
```suggestion
For development purposes, you can serve your files with a
```

File: docs/usage/modularizing.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
+
+(modularizing-python-codes)=
+# Modularizing Python codes
+
+Pyodide provides a simple API {any}`pyodide.runPython` to run Python code.
+However, when your Python codes grow bigger, putting hundreds of lines inside `runPython` is not scalable.
+
+The best way to run Python code with Pyodide is:
```
Comment (rth) [inline docs/usage/modularizing.md:8]:
```suggestion
For larger projects, the best way to run Python code with Pyodide is:
```

File: docs/usage/modularizing.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+1. write a Python package
```
Comment (rth) [inline docs/usage/modularizing.md:10]:
```suggestion
1. create a Python package
```

File: docs/usage/modularizing.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+pkg = pyodide.pyimport("script");
+pkg.do_something();
+```
+
+```pyodide
+// Archive
+await pyodide.runPythonAsync(`
+    from pyodide.http import pyfetch
+    response = await pyfetch("https://.../your_package.tar.gz")
+    await response.unpack_archive()
```
Comment (rth) [inline docs/usage/modularizing.md:57]:
```suggestion
    await response.unpack_archive()   # by default, unpacks to the current dir 
```

File: docs/usage/modularizing.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+pkg.do_something();
+```
+
+```{admonition} How can I prefetch package files before loading Pyodide
+:class: info
+
+If you want to prefetch package files before loading Pyodide,
+you can use JS `fetch` to download packages.
+After fetching packages, Pyodide provides {any}`pyodide.unpackArchive()` to save them
+into the Pyodide virtual file system.
```
Comment (rth) [inline docs/usage/modularizing.md:69]: I'm not convinced by this section. We should document how to preload packages indeed, however, I'm not sure that saying just use JS `fetch` before loading Pyodide is the right solution. I was hoping something like https://github.com/pyodide/pyodide/issues/1576 would work, but I couldn't confirm it.

For now maybe better to remove this section altogether, so it doesn't block merging of this PR. It can always be re-added in a follow up PR.

File: docs/usage/modularizing.md (hunk 5 above)
Comment (rth) [reply docs/usage/modularizing.md:3]: Or `Loading custom Python code` since I guess most of this is going to be about that, and this should be separate from bundling code in packages  ?

**Review by rth — COMMENTED** [review]
Thanks ! A few comments below.

Also IMO the `modularize.md` page currently has redundant information with explanations on how to load packages with micropip in https://pyodide.org/en/stable/usage/loading-packages.html#installing-packages-from-pypi There are also repetitions between the FAQ entry and this new page which we should reduce.

I think it could indeed make sense to have a separate page for this, but then,
 - loading-packages.md -> (already exists) explains how to load packages
 - loading-python-code.md -> (page added in this MR) explains how to load Python code, which is not included as a package. For loading packages, it should link to the first page instead of re-explaining it. We would also need to link to https://pyodide.org/en/stable/development/new-packages.html#determining-if-creating-a-pyodide-package-is-necessary on how to create wheels (or maybe better move that explanation in some separate section later)

For the newly added page, maybe we could have the following subsections,

1. Running external code directly

I think mentioning
```
pyodide.runPython(await (await fetch("https://some_url/...")).text());
```
is still useful, and maybe mentioning its limitations.  Or do we want to remove all mentions to it altogether.

2. Loading then importing Python code

the current content of the page in this PR

WDYT?

File: docs/development/contributing.md (hunk 12 above)
Comment (rth) [reply docs/development/contributing.md:39]: Yeah, we only follow semver because for 0.x.y each version can have breaking changes. Pretty much each Pyodide version has breaking changes. Let's not have this discussion in this PR, to avoid blocking it, there is https://github.com/pyodide/pyodide/issues/988 already.

File: docs/usage/modularizing.md (hunk 18 above)
Comment (ryanking13) [reply docs/usage/modularizing.md:69]: Okay, I'll remove it!

File: docs/usage/modularizing.md (hunk 5 above)
Comment (ryanking13) [reply docs/usage/modularizing.md:3]: I'll go with `Loading custom Python code`

**Comment by ryanking13** [comment]
@rth Reorganized `Loading custom Python code` section to remove redundant information. How does it look now?

https://pyodide--2070.org.readthedocs.build/en/2070/usage/loading-custom-python-code.html

File: docs/index.rst [hunk 19]
```diff
@@ -23,12 +23,12 @@ language, catch it in the other), async/await, and much more.
 
 When used inside a browser, Python has full access to the Web APIs.
 
-Try Pyodide (no installation needed)
+Try Pyodide
 ------------------------------------
```
Comment (rth) [inline docs/index.rst:27]:
```suggestion
-----------
```

File: docs/usage/downloading-and-deploying.md [hunk 20]
```diff
@@ -10,7 +10,7 @@ Pyodide packages, including the `pyodide.js` file, are available from the JsDeli
 
 | channel             | indexURL                                         | Comments                                                                                 | REPL                                               |
 | ------------------- | ------------------------------------------------ | ---------------------------------------------------------------------------------------- | -------------------------------------------------- |
-| Latest release      | `https://cdn.jsdelivr.net/pyodide/v0.18.0/full/` | Recommended, cached by the browser                                                       | [link](https://pyodide.org/en/stable/console.html) |
+| Latest release      | `https://cdn.jsdelivr.net/pyodide/v0.19.0/full/` | Recommended, cached by the browser                                                       | [link](https://pyodide.org/en/stable/console.html) |
```
Comment (rth) [inline docs/usage/downloading-and-deploying.md:13]: I think this still mentions the latest stable release which is fine. Once we bump the version we will also do it here, but we should do in a separate commit. Otherwise this link doesn't exist currently.
```suggestion
| Latest release      | `https://cdn.jsdelivr.net/pyodide/v0.18.0/full/` | Recommended, cached by the browser                                                       | [link](https://pyodide.org/en/stable/console.html) |
```

File: docs/usage/loading-custom-python-code.md [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+
+```{admonition} What is pyfetch?
+:class: info
+
+Pyodide provides {any}`pyodide.http.pyfetch`,
+which is a convenient wrapper of JavaScript `fetch`.
+See {ref}`load-external-files-in-pyodide` for more information.
+```
+
+### In Javascript side
```
Comment (rth) [inline docs/usage/loading-custom-python-code.md:66]:
```suggestion
### From JavaScript
```

File: docs/usage/loading-custom-python-code.md [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+:class: warning
+
+If your Python code contains C extensions,
+it needs to be built in a specialized way (See {ref}`new-packages`).
+```
+
+## Loading then importing Python code
+
+It is also possible to download and import Python code from an external source.
+We recommend you to serve whole code as an archive, instead of downloading each Python scripts.
```
Comment (rth) [inline docs/usage/loading-custom-python-code.md:31]:
```suggestion
We recommend that you serve all files in an archive, instead of individually downloading each Python script.
```

File: docs/usage/loading-custom-python-code.md [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+However, when your Python code grow bigger, putting hundreds of lines inside `runPython` is not scalable.
+
+For larger projects, the best way to run Python code with Pyodide is:
+
+1. create a Python package
+1. load your Python package into the Pyodide (Emscripten) virtual file system
+1. import the package with ``let mypkg = pyodide.pyimport("mypkgname")``
+1. call into your package with ``mypkg.some_api(some_args)``.
+
+## Using wheel packages
```
Comment (rth) [inline docs/usage/loading-custom-python-code.md:15]:
```suggestion
## Using wheels
```

File: docs/usage/loading-custom-python-code.md [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+
+The most recommended way of serving custom Python code is making it a wheel package.
```
Comment (rth) [inline docs/usage/loading-custom-python-code.md:17]:
```suggestion
The best way of serving custom Python code is making it a package in the wheel (.whl) format.
```

I think technically wheel is package format, https://packaging.python.org/en/latest/glossary/#term-Wheel and "wheel package" sounds a bit unexpected to me.

File: docs/usage/loading-custom-python-code.md [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+If your Python code contains C extensions,
+it needs to be built in a specialized way (See {ref}`new-packages`).
+```
+
+## Loading then importing Python code
+
+It is also possible to download and import Python code from an external source.
+We recommend you to serve whole code as an archive, instead of downloading each Python scripts.
+
+### Inside Pyodide
```
Comment (rth) [inline docs/usage/loading-custom-python-code.md:33]:
```suggestion
### From Python
```

**Review by rth — APPROVED** [review]
A few wording suggestions, otherwise LGTM. Thanks @ryanking13 !

**Comment by ryanking13** [comment]
@rth Thanks for the thorough review!

**Comment by rth** [comment]
Thanks!
